### PR TITLE
Create RangeUpdatePointQueries.java

### PR DESCRIPTION
**Pull request by Braj65**
Added Fenwick tree RangeUpdatePointQueries program

### CODE REVIEW

File: FenwickTree/RangeUpdatePointQueries.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+		int[] inArr={0,0,0,0,0,0};
+		//Store the fenwick tree
+		int[] feniTree=new int[inArr.length+1];
+		//Update range using method 2
+		d.updateRangeMeth2(inArr, 2, 4, 10);
+		//Point query method 2
+		System.out.println(d.getPointMeth2(inArr, 3));
+		//Mehth 3 call
+		//Construct fenwick tree 
+		d.contructFeniTree(inArr);
```
Comment (williamfiset) [inline FenwickTree/RangeUpdatePointQueries.java]: Why does constructFeniTree not use its return value? Should it not be:
```java
int[] fenwickTree = d.contructFeniTree(inArr);
```

File: FenwickTree/RangeUpdatePointQueries.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
+public class RangeUpdatePointQueries {
+	
+	public static void main(String[] args) {
+		RangeUpdatePointQueries d=new RangeUpdatePointQueries();
+		//input array
+		int[] inArr={0,0,0,0,0,0};
+		//Store the fenwick tree
+		int[] feniTree=new int[inArr.length+1];
+		//Update range using method 2
+		d.updateRangeMeth2(inArr, 2, 4, 10);
```
Comment (williamfiset) [inline FenwickTree/RangeUpdatePointQueries.java]: What is the difference between updateRange and updateRangeMeth2? When should one be used over the other? Furthermore, why is _inArr_ being passed into the method and not _feniTree_?

File: FenwickTree/RangeUpdatePointQueries.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+		//Point query method 2
+		System.out.println(d.getPointMeth2(inArr, 3));
```
Comment (williamfiset) [inline FenwickTree/RangeUpdatePointQueries.java]: Can you explain what the difference between getPoint and getPointMeth2? I noticed that getPoint has a logarithmic time complexity while getPointMeth2 has a linear complexity. getPointMeth2 seems impractical if it's to be so slow.

File: FenwickTree/RangeUpdatePointQueries.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
+public class RangeUpdatePointQueries {
```
Comment (williamfiset) [inline FenwickTree/RangeUpdatePointQueries.java]: It'd be great if you could provide tests for your Fenwick tree implementation.

File: FenwickTree/RangeUpdatePointQueries.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+			feniTree[i]+=val;
+			i=getNext(i);
+		}
+	}
+	
+	//this equation finds next node to be updated(not going to be 
+	//parent node but next sibling and next branch's parent).
+	//it finds the next node by addding one to the right most set bit.
+	public int getNext(int index){
+		return index+(index & -index);
```
Comment (williamfiset) [inline FenwickTree/RangeUpdatePointQueries.java]: I recommend refactoring to something like:
```java
return index + lsb(index)
```
where lsb is the least significant bit function. See the other [Fenwick tree](https://github.com/williamfiset/data-structures/blob/master/FenwickTree/FenwickTree.java) implementation in this folder.

File: FenwickTree/RangeUpdatePointQueries.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+			index=getParent(index);
+		}
+		return sum;
+	}
+	
+	//This equation finds parent of the current node by
+	//flipping the righ most set bit. It will keep on finding parent
+	//till it reaches its own parent 2^x node.
+	public int getParent(int index){
+		return index-(index & -index);
```
Comment (williamfiset) [inline FenwickTree/RangeUpdatePointQueries.java]: Use lsb function.

Comment (Braj65) [reply FenwickTree/RangeUpdatePointQueries.java]: Sure. Will follow the patterns you have used in Fenwick tree program.

File: FenwickTree/RangeUpdatePointQueries.java (hunk 5 above)
Comment (Braj65) [reply FenwickTree/RangeUpdatePointQueries.java]: Will do the modifications.

File: FenwickTree/RangeUpdatePointQueries.java (hunk 1 above)
Comment (Braj65) [reply FenwickTree/RangeUpdatePointQueries.java]: I have written both the linear and logarithmic methods operating on same input array. Refactoring will make it clear.

File: FenwickTree/RangeUpdatePointQueries.java (hunk 3 above)
Comment (Braj65) [reply FenwickTree/RangeUpdatePointQueries.java]: updateRangeMeth2 and getPointMeth2- Its complexity is linear. It updates the input array with update values. If we want to update position 2 to 4 of input array with 10, it adds inArr[2]+=10 and inArr[5]+=10. When we want the point value of lets say position 3. We add all the values from position 0 to 3. That returns value 10. If we want point value for position 5. We add all the values from position 0 to 5. That returns value 0.

File: FenwickTree/RangeUpdatePointQueries.java (hunk 2 above)
Comment (Braj65) [reply FenwickTree/RangeUpdatePointQueries.java]: updateRangeMeth2 updates the input array(it has linear complexity) and doesn't involves the fenwick tree, hence it is getting passed in the method. I will remove the linear time method, refactor the code and push it again in the branch.

File: FenwickTree/RangeUpdatePointQueries.java (hunk 4 above)
Comment (Braj65) [reply FenwickTree/RangeUpdatePointQueries.java]: I will add some tests based on tests you have for any other algorithms. I will push with all the changes below in a day.

**Comment by Braj65** [comment]
I have updated the files based on above comments. Please review the code once you have some time and let me know if any file needs any modification or if anything is confusing in it. I will rectify.

**Comment by williamfiset** [comment]
Nice work! I just tweaked some of the tests because they weren't actually being run. I should have told you that test file need to ends in *Test.java. I really should change that...

**Comment by Braj65** [comment]
Thanks. I have the understanding for Segment tree lazy propagation using array structure. I will create pull request for it some time.

**Pull request merged**